src/main/java/com/google/api/codegen/util/testing/TestValueGenerator.java

```diff
@@ -0,0 +1,45 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.util.testing;
+
+import com.google.api.codegen.util.Name;
+import com.google.api.tools.framework.model.TypeRef;
+
+import java.util.HashMap;
+
+/**
+ * A utility class used by the test generators which populates values for primitive fields.
+ *
+ * <p>The populated values will be unique (except for boolean values) and deterministic based on the
+ * populated fields and the sequence.
+ */
+public abstract class TestValueGenerator {
+  protected HashMap<Name, String> valueTable = new HashMap<>();
+
+  public String getAndStoreValue(TypeRef type, Name identifier) {
+    if (!valueTable.containsKey(identifier)) {
+      String value = generateValue(type, identifier);
+      while (type.getPrimitiveTypeName() != "bool" && valueTable.containsValue(value)) {
+        // If the value already exists regenerate using a deterministically different identifier.
+        identifier = identifier.join("1");
+        value = generateValue(type, identifier);
+      }
+      valueTable.put(identifier, generateValue(type, identifier));
+    }
+    return valueTable.get(identifier);
+  }
+
+  protected abstract String generateValue(TypeRef typeRef, Name identifier);
```

Comment:
I think this is a situation where abstract class can be used.  According to the wikipedia, we should choose composition over inheritance to achieve polymorphic behavior, which is not needed here. The functionality of this generator is rather simple where each language just needs to override the generateValue method.

From the perspective of complexity, the inheritance approach only requires one type of class (TestValueGenerator and its subclasses) but the composition approach requires two (TestValueGenerator, an composition interface and its implementing classes).